### PR TITLE
Full implementation of a 'validated activation flow'

### PR DESCRIPTION
**Pull request by micschk**
In short, by setting validated_activation_mode=true, users are required to enter an initial token right upon activating two-factor-authentication in the CMS. 2FA is only activated if this token validates, thereby validating that the user has correctly scanned and setup a login-device.

Deactivation requires the user's password as a check. @TODO: currently the password entered will be in cleartext.

```yml
_2fa\Extensions\Member:
  validated_activation_mode: true
```

This PR replaces PR #17, which was a first version of this functionality. This new PR implements a full CMS interface as opposed to #17, which required a custom implementation.
Some changes from PR #21 are necessarily duplicated in this PR, as this functionality builds further on PR #21.

Some screens, to give you an idea:
screen shot 2017-06-20 at 15 36 41
screen shot 2017-06-20 at 15 37 38
screen shot 2017-06-20 at 15 38 09
screen shot 2017-06-20 at 15 38 51
screen shot 2017-06-20 at 15 39 18
screen shot 2017-06-20 at 15 39 42

FIX: #22

### CODE REVIEW

File: _config/2fa.yml [hunk 1]
```diff
@@ -15,3 +15,11 @@ Injector:
     class: '_2fa\CMSProfileController'
 CMSSecurity:
   reauth_enabled: false
+
+LeftAndMain:
+  extra_requirements_css:
+    - silverstripe-twofactorauth/css/twofactorstyles.css
```
Comment (camfindlay) [inline _config/2fa.yml]: "installer-name": "twofactorauth", in composer.json so the path here should be `twofactorauth` and avoid duplicate.

File: _config/2fa.yml [hunk 2]
```diff
@@ -15,3 +15,11 @@ Injector:
...
+    - twofactorauth/css/twofactorstyles.css
+  extra_requirements_javascript:
+    - silverstripe-twofactorauth/js/twofactorinteractions.js
+    - twofactorauth/js/twofactorinteractions.js
```
Comment (camfindlay) [inline _config/2fa.yml]: Same as a above.

File: code/extension/Member.php [hunk 3]
```diff
@@ -88,6 +95,12 @@ public function OTPSeed()
      */
     public function updateCMSFields(\FieldList $fields)
     {
+        // Generate default token (allows scanning the QR at the moment of activation and (optionally) validate before activating 2FA)
+        if(!$this->owner->TOTPToken && self::validated_activation_mode()) {
+            $this->generateTOTPToken();
```
Comment (camfindlay) [inline code/extension/Member.php:100]: $this->owner?

File: code/form/LoginForm.php [hunk 4]
```diff
@@ -7,6 +7,16 @@
 
 class LoginForm extends \MemberLoginForm
 {
+    private static $allowed_actions = array('cancel');
+
+    /**
+     * Action to unset the TOTP.ID session var to allow going back to the normal (email/pw) login form
+     */
+    public function cancel() {
```
Comment (camfindlay) [inline code/form/LoginForm.php:15]: Isn't this already implemented and merged to master?

**Comment by camfindlay** [comment]
Config docs too please :shipit:

File: _config/2fa.yml (hunk 1 above)
Comment (micschk) [reply _config/2fa.yml]: Fixing now

File: _config/2fa.yml (hunk 2 above)
Comment (micschk) [reply _config/2fa.yml]: Fixing now

File: code/extension/Member.php (hunk 3 above)
Comment (micschk) [reply code/extension/Member.php:100]: Yes, it's applied as a DataExtension (or am I missing your point?)

File: code/form/LoginForm.php (hunk 4 above)
Comment (micschk) [reply code/form/LoginForm.php:15]: Yes, it's included again here because of my Git incompetency. I'll try & remove it.

Comment (micschk) [reply code/form/LoginForm.php:15]: Sorry, cannot seem to remove it from this commit... Some help appreciated.

**Comment by micschk** [comment]
Should be all done now...

**Comment by camfindlay** [comment]
Nice work, just testing it locally at the moment :+1:

**Comment by camfindlay** [comment]
Manual testing... super happy with the workflow. 

LGTM

**Comment by camfindlay** [comment]
Seems to be non-breaking with the way the config is false by default (I'd go so far as to suggest we look to switch the default to true on a next major release).

**Comment by micschk** [comment]
Great! Thanks for merging!

Maybe it'd be even better to phase out the previous flow in a next release? It does require a bit of overhead in the amount of code to maintain, and (in my opinion) the new flow delivers a better user experience in any scenario I can think of.

**Comment by camfindlay** [comment]
Was going to suggest this also. Could be a 3.0.0 (given it changes default behaviour and we may well strip out some code) which leaves us a 4.0.0 for SS4 compat work (the numbers line up! 🎉 )

**Pull request merged**